File: environment/src/main/java/com/chutneytesting/environment/domain/EnvironmentService.java
/*
 * Copyright 2017-2023 Enedis
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.chutneytesting.environment.domain;

import com.chutneytesting.environment.domain.exception.AlreadyExistingEnvironmentException;
import com.chutneytesting.environment.domain.exception.AlreadyExistingTargetException;
import com.chutneytesting.environment.domain.exception.CannotDeleteEnvironmentException;
import com.chutneytesting.environment.domain.exception.EnvironmentNotFoundException;
import com.chutneytesting.environment.domain.exception.InvalidEnvironmentNameException;
import com.chutneytesting.environment.domain.exception.TargetNotFoundException;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class EnvironmentService {

    private static final String NAME_VALIDATION_REGEX = "[a-zA-Z0-9_\\-]{3,20}";
    private static final Pattern NAME_VALIDATION_PATTERN = Pattern.compile(NAME_VALIDATION_REGEX);

    private final Logger logger = LoggerFactory.getLogger(EnvironmentService.class);
    private final EnvironmentRepository environmentRepository;

    public EnvironmentService(EnvironmentRepository environmentRepository) {
        this.environmentRepository = environmentRepository;
    }

    public Set<String> listEnvironmentsNames() {
        return new LinkedHashSet<>(environmentRepository.listNames());
    }

    public Set<Environment> listEnvironments() {
        return environmentRepository
            .listNames()
            .stream()
            .map(environmentRepository::findByName)
            .collect(Collectors.toSet());
    }

    public Environment createEnvironment(Environment environment) throws InvalidEnvironmentNameException, AlreadyExistingEnvironmentException {
        return createEnvironment(environment, false);
    }

    public Environment createEnvironment(Environment environment, boolean force) throws InvalidEnvironmentNameException, AlreadyExistingEnvironmentException {
        if (!force && envAlreadyExist(environment)) {
            throw new AlreadyExistingEnvironmentException("Environment [" + environment.name + "] already exists");
        }
        createOrUpdate(environment);
        return environment;
    }

    public Environment getEnvironment(String environmentName) throws EnvironmentNotFoundException {
        return environmentRepository.findByName(environmentName);
    }

    public void deleteEnvironment(String environmentName) throws EnvironmentNotFoundException, CannotDeleteEnvironmentException {
        environmentRepository.delete(environmentName);
    }

    public void updateEnvironment(String environmentName, Environment newVersion) throws InvalidEnvironmentNameException, EnvironmentNotFoundException {
        Environment previousEnvironment = environmentRepository.findByName(environmentName);
        Environment newEnvironment = Environment.builder()
            .from(previousEnvironment)
            .withName(newVersion.name)
            .withDescription(newVersion.description)
            .build();
        createOrUpdate(newEnvironment);
        if (!newEnvironment.name.equals(environmentName)) {
            environmentRepository.delete(environmentName);
        }
    }

    public List<Target> listTargets(TargetFilter filters) {
        Set<Target> targets;
        if (filters != null && StringUtils.isNotBlank(filters.environment())) {
            targets = environmentRepository.findByName(filters.environment()).targets;
        } else {
            targets = listEnvironments()
                .stream()
                .flatMap(environment -> environment.targets.stream()).collect(Collectors.toSet());
        }
        return targets
            .stream()
            .filter(target -> match(target, filters))
            .collect(Collectors.toList());
    }


    public Set<String> listTargetsNames() {
        return listEnvironments().stream()
            .flatMap(environment -> environment.targets.stream().map(target -> target.name))
            .collect(Collectors.toSet());

    }

    public Target getTarget(String environmentName, String targetName) {
        Environment environment = environmentRepository.findByName(environmentName);
        return environment.getTarget(targetName);
    }

    public void addTarget(Target target) throws EnvironmentNotFoundException, AlreadyExistingTargetException {
        Environment environment = environmentRepository.findByName(target.environment);
        Environment newEnvironment = environment.addTarget(target);
        createOrUpdate(newEnvironment);
    }

    public void deleteTarget(String environmentName, String targetName) throws EnvironmentNotFoundException, TargetNotFoundException {
        Environment environment = environmentRepository.findByName(environmentName);
        Environment newEnvironment = environment.deleteTarget(targetName);
        createOrUpdate(newEnvironment);
    }

    public void deleteTarget(String targetName) throws EnvironmentNotFoundException, TargetNotFoundException {
        environmentRepository.getEnvironments()
            .stream()
            .filter(env -> env.targets.stream().map(target -> target.name).toList().contains(targetName))
            .forEach(env -> {
                Environment newEnvironment = env.deleteTarget(targetName);
                createOrUpdate(newEnvironment);
            });
    }

    public void updateTarget(String previousTargetName, Target targetToUpdate) throws EnvironmentNotFoundException, TargetNotFoundException {
        Environment environment = environmentRepository.findByName(targetToUpdate.environment);
        Environment newEnvironment = environment.updateTarget(previousTargetName, targetToUpdate);
        createOrUpdate(newEnvironment);
        logger.debug("Updated target " + previousTargetName + " as " + targetToUpdate.name);
    }

    private void createOrUpdate(Environment environment) {
        if (!NAME_VALIDATION_PATTERN.matcher(environment.name).matches()) {
            throw new InvalidEnvironmentNameException("Environment name must be of 3 to 20 letters, digits, underscore or hyphen");
        }
        environmentRepository.save(environment);
    }

    private boolean envAlreadyExist(Environment environment) {
        return environmentRepository.listNames().stream().map(String::toUpperCase)
            .toList().contains(environment.name.toUpperCase());
    }

    private boolean match(Target target, TargetFilter filters) {
        if (filters == null) {
            return true;
        }

        boolean matchName = true;
        if (StringUtils.isNotBlank(filters.name())) {
            matchName = filters.name().equals(target.name);
        }

        return matchName;
    }
}

File: ui/src/app/core/services/environment.service.ts
/**
 * Copyright 2017-2023 Enedis
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Environment, EnvironmentVariable, Target, TargetFilter } from '@model';
import { environment as server } from '../../../environments/environment';
import { HttpClient, HttpContext, HttpHeaders, HttpParams } from '@angular/common/http';
import { map, tap } from 'rxjs/operators';
import { saveAs } from 'file-saver';
import { FileSaverService } from 'ngx-filesaver';

@Injectable({
    providedIn: 'root'
})
export class EnvironmentService {

    private envBaseUrl = '/api/v2/environments';
    private targetBaseUrl = '/api/v2/targets';
    private variablesBaseUrl = '/api/v2/variables';

    constructor(private http: HttpClient,
                private fileSaverService: FileSaverService) {
    }

    list(): Observable<Environment[]> {
        return this.http
            .get<Environment[]>(server.backend + this.envBaseUrl)
            .pipe(
                map(res => res.sort((t1, t2) => t1.name.toUpperCase() > t2.name.toUpperCase() ? 1 : 0))
            );
    }

    names(): Observable<Array<string>> {
        return this.http.get<Array<string>>(server.backend + this.envBaseUrl + '/names');
    }

    get(environmentName: string): Observable<Environment> {
        return this.http.get<Environment>(server.backend + this.envBaseUrl + '/' + environmentName);
    }

    export(environmentName: string): Observable<Environment> {
        const fileName = `env.${environmentName}.chutney.json`;
        return this.get(environmentName).pipe(
            tap(env => {
                this.fileSaverService.saveText(JSON.stringify(env), fileName)
            }),
        );
    }

    import(file: File): Observable<Environment> {
        const formData = new FormData();
        formData.append('file', file);
        return this.http.post<Environment>(server.backend + this.envBaseUrl, formData);
    }

    create(environment: Environment): Observable<Object> {
        return this.http.post(server.backend + this.envBaseUrl, environment);
    }

    delete(environmentName: string): Observable<Object> {
        return this.http.delete(server.backend + this.envBaseUrl + '/' + environmentName);
    }

    update(environmentName: string, environment: Environment): Observable<Object> {
        return this.http.put(server.backend + this.envBaseUrl + '/' + environmentName, environment);
    }


    environmentTarget(environmentName: string, targetName: string): Observable<Target> {
        return this.http.get<Target>(server.backend + this.envBaseUrl + '/' + environmentName + '/targets/' + targetName);
    }

    getTargets(filters: TargetFilter): Observable<Target[]> {
        const params = Object.entries(filters)
            .filter(([key, value]) => value != null)
            .reduce((params, [key, value]) => params.append(key, value), new HttpParams());
        return this.http.get<Target[]>(server.backend + this.targetBaseUrl, {params});
    }

    updateTarget(targetName: string, target: Target): Observable<Object> {
        return this.http.put(server.backend + this.targetBaseUrl + '/' + targetName, target);
    }

    addTarget(target: Target): Observable<Object> {
        return this.http.post(server.backend + this.targetBaseUrl, target);
    }

    deleteTarget(targetName: string): Observable<void> {
        return this.http.delete<void>(server.backend + this.targetBaseUrl + '/' + targetName);
    }

    deleteEnvironmentTarget(environmentName: string, targetName: string): Observable<Object> {
        return this.http.delete(server.backend + this.envBaseUrl + '/' + environmentName + '/targets/' + targetName);
    }

    exportTargetOn(environmentName: string, targetName: string): Observable<void> {
        return this.environmentTarget(environmentName, targetName)
            .pipe(
                map(target => {
                    delete target.environment;
                    const fileName = `${environmentName}-${targetName}.chutney.json`;
                    this.fileSaverService.saveText(JSON.stringify(target), fileName);
                    return;
                })
            );
    }

    importTarget(file: File, environment: string): Observable<Target> {
        const formData = new FormData();
        formData.append('file', file);
        return this.http.post<Target>(server.backend + this.envBaseUrl + '/' + environment + '/targets/', formData);
    }


    addVariable(variables: EnvironmentVariable[]): Observable<void> {
        return this.http.post<void>(server.backend + this.variablesBaseUrl, variables);
    }
    updateVariable(key: string, values: EnvironmentVariable[]): Observable<void> {
        return this.http.put<void>(server.backend + this.variablesBaseUrl + '/' + key, values);
    }

    deleteVariable(key: string): Observable<void> {
        return this.http.delete<void>(server.backend + this.variablesBaseUrl + '/' + key);
    }

}
